src/controllers/smartContractController.ts

```diff
@@ -0,0 +1,129 @@
+            abi,
+        });
+        this.transactionWatcher = new TransactionWatcher(new ProviderWrapper(networkProvider));
+        this.txComputer = new TransactionComputer();
+    }
+
+    async createTransactionForDeploy(sender: IAccount, options: ContractDepoyInput): Promise<Transaction> {
+        const transaction = this.factory.createTransactionForDeploy({ ...options, sender: sender.address });
+
+        transaction.nonce = options.nonce;
+        transaction.signature = await sender.sign(this.txComputer.computeBytesForSigning(transaction));
+
+        return transaction;
+    }
+
+    async awaitCompletedDeploy(txHash: string): Promise<SmartContractDeployOutcome> {
+        const transaction = await this.transactionWatcher.awaitCompleted(txHash);
+        return this.parseDeploy(transaction);
+    }
+
+    async createTransactionForUpgrade(sender: IAccount, options: ContractUpgradeInput): Promise<Transaction> {
+        const transaction = this.factory.createTransactionForUpgrade({ ...options, sender: sender.address });
+
+        transaction.nonce = options.nonce;
+        transaction.signature = await sender.sign(this.txComputer.computeBytesForSigning(transaction));
+
+        return transaction;
+    }
+
+    async createTransactionForExecute(sender: IAccount, options: TransactionInput): Promise<Transaction> {
+        const transaction = this.factory.createTransactionForExecute({ ...options, sender: sender.address });
+
+        transaction.nonce = options.nonce;
+        transaction.signature = await sender.sign(this.txComputer.computeBytesForSigning(transaction));
+
+        return transaction;
+    }
+
+    queryContract(contract: IAddress, func: string, args: any[], caller?: IAddress, value?: bigint): Promise<any[]> {
+        return this.queryController.query({
+            contract: contract.bech32(),
+            function: func,
+            arguments: args,
+            caller: caller ? caller.bech32() : undefined,
+            value: BigInt(value ?? 0),
+        });
+    }
+
+    parseDeploy(transactionOnNetwork: ITransactionOnNetwork): SmartContractDeployOutcome {
+        return this.parser.parseDeploy({ transactionOnNetwork });
+    }
+}
+
+type ContractDepoyInput = {
+    nonce: bigint;
+    bytecode: Uint8Array;
+    gasLimit: bigint;
+    arguments: any[];
+    nativeTransferAmount?: bigint;
+    isUpgradeable?: boolean;
+    isReadable?: boolean;
+    isPayable?: boolean;
+    isPayableBySmartContract?: boolean;
+};
+
+type TransactionInput = {
+    nonce: bigint;
+    contract: IAddress;
+    gasLimit: bigint;
+    function: string;
+    arguments?: any[];
+    nativeTransferAmount?: bigint;
+    tokenTransfers?: TokenTransfer[];
+};
+
+type ContractUpgradeInput = ContractDepoyInput & { contract: IAddress };
+
+interface SmartContractDeployOutcome {
+    returnCode: string;
+    returnMessage: string;
+    contracts: DeployedSmartContract[];
+}
+
+class DeployedSmartContract {
```

Comment:
Tricky if it's defined as class. Maybe have it defined somewhere else, and used in the outcome parser itself (create instances and such)?

Tricky because: it's a class, and people might expect that `outcome.contracts` really are objects having `DeployedSmartContract` as prototype. But they won't have it (at runtime), they will be mere, plain objects. Maybe have it as type?

Thus, a few possible options:
1. have it as type instead of class
2. in the outcome parser, have returned objects be `DeployedSmartContract`
3. for deploy, re-map mere objects returned by the outcome parser to instances of your class
4. something else